docs: rewrite GitHub integration page as step-by-step tutorial (#2512)
apify/apify-docs: sources/platform/integrations/programming/github.md

---
title: GitHub integration
description: Connect Apify with GitHub to build Actors from a repository, rebuild on every push, and create issues automatically when an Actor run fails.
sidebar_label: GitHub
sidebar_position: 3
slug: /integrations/github
---

With the Apify integration for [GitHub](https://github.com/), you can create an Actor from a public or private repository, rebuild it automatically on every push, and trigger workflows in your repo when an Actor run fails, succeeds, or times out. To run automated tests and multi-branch builds with GitHub Actions, see [Continuous integration for Actors](/platform/actors/development/deployment/continuous-integration).

## Prerequisites

To use the GitHub integration, you need:

- An [Apify account](https://console.apify.com/).
- A GitHub account with access to the repository you want to link.

> 

Reviewer: I think this should be moved below ## Create an Actor from a GitHub repository heading. Right now it feels out of place before it. Especially with another `Prerequisites` section being actually below?

### Step 2: Connect your GitHub account

Under **Link a Git repository**, click **GitHub**. Follow the prompts on github.com to authorize Apify. You can grant access to your personal account, an organization, or specific repositories.

![Authorize Apify on GitHub](../images/apify-git-repository-add.png)

To switch between authorized users and organizations, use the account dropdown.

![Account picker](../images/apify-git-repository-account.png)

### Step 3: Select a repository

Pick the repository you want to link. If you don't see it, use the **Search** field to find it by name.

![Repository search](../images/apify-git-repository-search.png)

Apify creates the Actor as soon as you select a repository, links its source to the repository, and uses the default branch unless you change it in the Actor's **Source** settings.

:::tip Private repositories

For private repositories, configure a [deployment key](/platform/actors/development/deployment/source-types#private-repositories) so Apify can clone the code.

:::

## Build automatically on every push

After you link an Actor to a GitHub repository, add a webhook in GitHub to trigger a new build on every push:

> 1. In Apify Console, open the Actor's **API** tab and select **API Endpoints**. Copy the **Build Actor** endpoint URL. It has this format:

Reviewer: Is this a tab or a dropdown?

<img width="411" height="370" alt="Image" src="https://github.com/user-attachments/assets/bdcda85b-8482-4090-ad42-06cc73fc29c2" />

1. In GitHub, open **Settings > Developer settings > Personal access tokens > Fine-grained tokens** and click **Generate new token**.
1. Set **Repository access** to **Only select repositories** and pick the repository where you want issues to be created.
1. Under **Repository permissions**, set **Issues** to **Read and write**.
1. Generate the token and copy it. You'll paste it into the webhook headers in Step 3.

:::warning Treat the token as a secret

Anyone with this token can create issues in the selected repository. Don't commit it or share it in screenshots.

:::

### Step 2: Add a webhook on the Actor

1. In Apify Console, open the Actor and go to the **Integrations** tab.

    ![Add integration page on an Actor's Integrations tab](../images/integrations-tab.png)

1. Under **Connect with Apify**, click **HTTP webhook**.
1. Configure the webhook:
    - **Event types**: select `Run failed` (`ACTOR.RUN.FAILED`).
    - **URL**: `https://api.github.com/repos/OWNER/REPO/issues`, replacing `OWNER` and `REPO` with your repository details.

### Step 3: Set the headers and payload

In the same webhook form, configure the request that GitHub expects.

> Set **Headers template** to authenticate with the personal access token and tell GitHub which API version to use:

Reviewer: Is `tell` the best word here? It anthropomorphizes the API. `specify` seems more accurate.

```json
{
    "Authorization": "Bearer YOUR_GITHUB_TOKEN",
    "Accept": "application/vnd.github+json",
    "X-GitHub-Api-Version": "2022-11-28"
}
```

Enable **Interpolate variables in string fields**, then set **Payload template** so the issue title and body include the failed run details:

```json
{
    "title": "Actor run {{resource.id}} failed",
    "body": "Actor [{{resource.actId}}](https://console.apify.com/actors/{{resource.actId}}) finished with status `{{resource.status}}`.\n\nRun: https://console.apify.com/actors/{{resource.actId}}/runs/{{resource.id}}\nStarted: {{resource.startedAt}}\nFinished: {{resource.finishedAt}}\nExit code: {{resource.exitCode}}",
    "labels": ["actor-failure"]
}
```

For the full list of variables you can use, see [Webhook actions](/platform/integrations/webhooks/actions#available-variables).

### Step 4: Save and test the webhook

> 1. Click **Save** to add the webhook.

Reviewer: Tangent, not actionable for this PR: we should rethink our use of click more broadly. 

Either prefer select outright, or formalize the distinction (click for buttons, select for choices in a list).

Author: Added as a topic for the sync next week

1. Click **Test** to send a sample payload to the GitHub API. Verify a new issue appears in the repository.

> 1. Trigger a real failure (for example, by aborting a run that's configured to fail or by running the Actor with invalid input) and confirm an issue is created.

Reviewer: Aborting =/= failing. It fires `ACTOR.RUN.ABORTED`, and this webhook is scoped for `ACTOR.RUN.FAILED`, so abort won't trigger it. Also, `a run that's configured to fail` reads oddly. Runs aren't configured to fail, Actors fail given certain inputs.

```suggestion
1. Trigger a real failure (for example, run the Actor with input that you know will cause a failure) and confirm an issue is created.
```

---
title: GitHub integration
description: Connect Apify with GitHub to build Actors from a repository, rebuild on every push, and create issues automatically when an Actor run fails.
sidebar_label: GitHub
sidebar_position: 3
slug: /integrations/github
---

import ThirdPartyDisclaimer from '@site/sources/_partials/_third-party-integration.mdx';

With the Apify integration for [GitHub](https://github.com/), you can create an Actor from a public or private repository, rebuild it automatically on every push, and trigger workflows in your repo when an Actor run fails, succeeds, or times out. To run automated tests and multi-branch builds with GitHub Actions, see [Continuous integration for Actors](/platform/actors/development/deployment/continuous-integration).

> <ThirdPartyDisclaimer />

Reviewer: I'll do a more thorough review later today, but wanted to ask first. Why add the third-party disclaimer here? The other pages in `programming/` don't carry it.

![Let's build a new Actor page in Apify Console](../images/apify-git-repository.png)

### Step 2: Connect your GitHub account

Under **Link a Git repository**, click **GitHub**. Follow the prompts on github.com to authorize Apify. You can grant access to your personal account, an organization, or specific repositories.

![Authorize Apify on GitHub](../images/apify-git-repository-add.png)

To switch between authorized users and organizations, use the account dropdown.

![Account picker](../images/apify-git-repository-account.png)

### Step 3: Select a repository

Pick the repository you want to link. If you don't see it, use the **Search** field to find it by name.

![Repository search](../images/apify-git-repository-search.png)

Apify creates the Actor as soon as you select a repository, links its source to the repository, and uses the default branch unless you change it in the Actor's **Source** settings.

:::tip Private repositories

For private repositories, configure a [deployment key](/platform/actors/development/deployment/source-types#private-repositories) so Apify can clone the code.

:::

## Build automatically on every push

> {/* REVIEWER QUESTION: The "Link a Git repository" panel says "set Apify to automatically build the Actor on changes", which suggests the native GitHub flow above already wires up auto-builds on push. If so, the manual GitHub webhook recipe below is now the legacy/advanced path. Please confirm the current behavior and let me know whether to (a) keep the manual recipe as-is, (b) replace it with a one-line note + link to CI for Actors, or (c) keep both with a short "the flow above already triggers builds; use this for finer control" intro. */}

Reviewer 2: No, unfortunately, right now, it does not work that way. You still need to setup webhook manually. We are working on improvements, but for now, let's describe it correctly -> so (a) keep the manual recipe as-is.
